### PR TITLE
game_list: Mark games as favorite to make them appear at the top.

### PR DESCRIPTION
**Pull request by Kewlan**
Useful if you have a lot of games!

image
image

I decided to have it store the program_ids instead of the full path. The benefit is that games can be moved without having to re-add them, plus I'm guessing most are not using homebrew.

All star icons are from this one: https://icons8.com/icon/8uAtuJQJ4jhd/star-filled

### CODE REVIEW

File: src/yuzu/game_list.h [hunk 1]
```diff
@@ -112,10 +112,15 @@ private slots:
 
     void RefreshGameDirectory();
 
+    void ToggleFavorite(u64 program_id);
+    void AddFavorite(const u64& program_id);
```
Comment (lioncash) [inline src/yuzu/game_list.h:116]: These don't need to be const references

File: src/yuzu/game_list.cpp [hunk 2]
```diff
@@ -84,6 +85,10 @@ void GameListSearchField::setFilterResult(int visible, int total) {
     label_filter_result->setText(tr("%1 of %n result(s)", "", total).arg(visible));
 }
 
+bool GameListSearchField::SearchFieldEmpty() {
```
Comment (lioncash) [inline src/yuzu/game_list.cpp:88]:
```suggestion
bool GameListSearchField::SearchFieldEmpty() const {
```

File: src/yuzu/game_list.cpp [hunk 3]
```diff
@@ -427,6 +442,13 @@ void GameList::DonePopulating(const QStringList& watch_list) {
     emit ShowList(!IsEmpty());
 
     item_model->invisibleRootItem()->appendRow(new GameListAddDir());
+    item_model->invisibleRootItem()->insertRow(0, new GameListFavorites());
+    tree_view->setRowHidden(0, item_model->invisibleRootItem()->index(),
+                            UISettings::values.favorited_ids.size() == 0);
+    tree_view->expand(item_model->invisibleRootItem()->child(0)->index());
+    for (u64 id : UISettings::values.favorited_ids) {
```
Comment (Morph1984) [inline src/yuzu/game_list.cpp:449]: ditto with the other instance below
```suggestion
    for (const auto id : UISettings::values.favorited_ids) {
```

File: src/yuzu/game_list.cpp [hunk 4]
```diff
@@ -728,6 +771,57 @@ void GameList::RefreshGameDirectory() {
...
+    }
+}
+
+void GameList::AddFavorite(u64 program_id) {
+    const auto favorites_row = item_model->item(0);
+
+    for (int i = 1; i < item_model->rowCount() - 1; i++) {
+        const auto folder = item_model->item(i);
+        for (int j = 0; j < folder->rowCount(); j++) {
+            if (folder->child(j)->data(GameListItemPath::ProgramIdRole) == program_id) {
```
Comment (Morph1984) [inline src/yuzu/game_list.cpp:797]: ditto with the other instance below
```suggestion
            if (folder->child(j)->data(GameListItemPath::ProgramIdRole).toULongLong() == program_id) {
```

File: src/yuzu/game_list_p.h [hunk 5]
```diff
@@ -29,7 +29,8 @@ enum class GameListItemType {
     SdmcDir = QStandardItem::UserType + 3,
     UserNandDir = QStandardItem::UserType + 4,
     SysNandDir = QStandardItem::UserType + 5,
-    AddDir = QStandardItem::UserType + 6
+    AddDir = QStandardItem::UserType + 6,
+    Favorites = QStandardItem::UserType + 7
```
Comment (Morph1984) [inline src/yuzu/game_list_p.h:33]:
```suggestion
    Favorites = QStandardItem::UserType + 7,
```

File: src/yuzu/game_list.cpp [hunk 6]
```diff
@@ -728,6 +771,57 @@ void GameList::RefreshGameDirectory() {
...
+        UISettings::values.favorited_ids.removeOne(program_id);
+        RemoveFavorite(program_id);
+        if (UISettings::values.favorited_ids.size() == 0) {
+            tree_view->setRowHidden(0, item_model->invisibleRootItem()->index(), true);
+        }
+    }
+}
+
+void GameList::AddFavorite(u64 program_id) {
+    const auto favorites_row = item_model->item(0);
```
Comment (Morph1984) [inline src/yuzu/game_list.cpp:792]: ditto with any other instances below
```suggestion
    const auto* favorites_row = item_model->item(0);
```

File: src/yuzu/game_list.cpp [hunk 7]
```diff
@@ -728,6 +771,57 @@ void GameList::RefreshGameDirectory() {
...
+
+    for (int i = 1; i < item_model->rowCount() - 1; i++) {
+        const auto folder = item_model->item(i);
```
Comment (Morph1984) [inline src/yuzu/game_list.cpp:795]:
```suggestion
        const auto* folder = item_model->item(i);
```

**Review by Morph1984 — COMMENTED** [review]
Looks really good, some minor nits

File: src/yuzu/game_list.cpp (hunk 6 above)
Comment (Kewlan) [reply src/yuzu/game_list.cpp]: I also removed `const` because `appendRow()` is used below.

**Review by bunnei — APPROVED** [review]

**Comment by bunnei** [comment]
@Kewlan great change! Thank you!

**Pull request merged**